### DISCUSSION TITLE
Use of Repositories from the HTTP routes

### DISCUSSION THREAD

**Discussion by yobuntu** [pycabook/pycabook#34] [Q&A]
Would it break the clean architecture desing to set the repository globally for the whole HTTP app ?

I'm a bit bothered by instancing a new repository in each HTTP route for the following reasons:

- DRY: the repository instantiation will be done in each and every route
- It make the use of route decorator a bit weird as, if my understanding is correct, the decorator would need to instantiate his own repository if in need access to stored data (stuff like user, permissions that one does not want recode in each route) (writing this make me think this one may not be valid as in clean design, maybe the permissions are to be managed at domain level…)

anyway i'd be glad to have your insight about this.

**Comment by lgiordani** [comment]
Oh, yes, I agree. It _might_ be better, as in "definitely performing better". I think this is one of those decisions that are half in the real of architecture and half in the real of performances. I would however avoid to have a giant repository that encompasses all use cases, as this would make use cases less modular, and ultimately couple some parts of the system that could be kept separate. Anyway, where to trace the line depends on the application and on the workload. A good approach to get the best of both worlds might be to leverage cached instances of the repository. Does it make sense?

**Comment by yobuntu** [comment]
> I would however avoid to have a giant repository that encompasses all use cases, as this would make use cases less modular, and ultimately couple some parts of the system that could be kept separate

I'm glad i asked this question, as  was totally missing the point on separate repositories !

I may need to read some literature and seriously think about how to split a single relational database in separates repositories (as the name implies there are a lot of relations between tables that are in the same database.

Thank you for all the responses you addressed yesterday beside this one btw.

I really look forward to read the second edition of your book.

Cheers
